File: Ex02/Animals_ObjC/Snake.m
//
//  Snake.m
//  Animals
//
//  Created by Elisabeth Hoppe on 22.06.15.
//  Copyright (c) 2015 Elisabeth Hoppe. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "Snake.h"

@implementation Snake:Reptile
- (id) initWithParams: (NSMutableString*)name_ carnivore:(BOOL)carnivore_ bodyLength:(int)bodyLength_
{
    self = [super initWithParams:name_ carnivore:carnivore_];
    [self setBodyLength:bodyLength_];
    return self;
}

-(void) setBodyLength:(int)bodyLength_
{
    bodyLength = bodyLength_;
}

-(int)getBodyLength
{
    return bodyLength;
}

-(void)toString
{
    //TODO;
}

-(BOOL)equals:(Snake *)snake
{
    
    
    //check the super class equality
    BOOL reptileEqualiy = [super equals:self];
    
    if(reptileEqualiy == YES)
    {
        //check for equality in own class
        if(bodyLength == [snake getBodyLength])
        {
            return YES;
        }
        else
        {
            return NO;
        }
    }
    else
    {
        return NO;
    }
    
    return NO;
}

@end

File: Ex02/Animals_JS/js/Snake.js
var Snake = function(name, carnivore, bodyTemperature){
	//calling super constructor
	Reptile.call(this, name, carnivore, bodyTemperature);
	
	//own (subclass) attributes
	this.setBodyLength = function(bodyLength){
		this.pBodyLength = bodyLength;
	};
	
	this.getBodyLength = function(){
		return this.pBodyLength;
	};
	
};

Snake.prototype = Object.create(Reptile.prototype);
Snake.prototype.constructor = Snake;

////// IMPLEMENTATION OF ABSTRACT METHODS
Snake.prototype.toString = function (){
	
	//TODO: nachfragen nach was alphabetisch ordnen (Member Name wie z.B. carnivore oder Wert z.B. true bei carnivore)
	console.log(this.getName()+"   "+ this.isCarnivore() + " "+ this.getBodyTemperature() 
		+ " " + this.getBodyLength());
	
}

Snake.prototype.equals = function(snakeOb){
	
	var reptileEqality = Object.getPrototypeOf(Snake.prototype).equals(snakeOb);
	
	if(reptileEquality == true)
	{
		if( (this.getBodyLength() == snakeOb.getBodyLength()) )
		{
			return true;
		}
		return false;
	}
	
	else
	{
		return false;
	}
	
	
 }


function createSnake() {
	try{
		console.log("creating snake...");
		var snake1 = new Snake ("Snake_1" , "true", 23.4);
		snake1.setBodyLength(24.5);
		snake1.toString();
		var snake2 = new Snake("Snake_2", "true", 24.6);
		snake2.setBodyLength(42.7);
		snake2.toString();
	}
	catch(e){
		console.log("Exception:" + e);
		return e;
	}
};
